cpp/src/Ice/ObjectAdapterFactory.cpp

```diff
@@ -39,59 +37,116 @@ IceInternal::ObjectAdapterFactory::shutdown()
-        adapters.end(),
-        [](const shared_ptr<ObjectAdapterI>& adapter) { adapter->deactivate(); });
+    for (const auto& adapter : adapters)
+    {
+        adapter->deactivate();
+    }
 }
 
 void
-IceInternal::ObjectAdapterFactory::waitForShutdown()
+IceInternal::ObjectAdapterFactory::waitForShutdown() noexcept
 {
     list<shared_ptr<ObjectAdapterI>> adapters;
-
     {
         unique_lock lock(_mutex);
+        if (_shutdownCompleted)
+        {
+            return;
+        }
 
-        //
         // First we wait for the shutdown of the factory itself.
-        //
         _conditionVariable.wait(lock, [this] { return !_instance; });
         adapters = _adapters;
     }
 
     // Now we wait for deactivation of each object adapter.
-    for_each(
-        adapters.begin(),
-        adapters.end(),
-        [](const shared_ptr<ObjectAdapterI>& adapter) { adapter->waitForDeactivate(); });
+    for (const auto& adapter : adapters)
+    {
+        adapter->waitForDeactivate();
+    }
+
+    {
+        lock_guard lock(_mutex);
+        _shutdownCompleted = true;
+    }
+}
+
+void
+IceInternal::ObjectAdapterFactory::waitForShutdownAsync(function<void()> completed) noexcept
+{
+    assert(completed); // the caller (Communicator) makes sure completed is callable.
+
+    bool executeCallback = false;
+    {
+        lock_guard lock(_mutex);
+        if (_shutdownCompleted)
+        {
+            executeCallback = true; // execute outside the lock
+        }
+        else
+        {
+            // Start the shutdown completed thread if there is no callback yet.
+            if (_shutdownCompletedCallbacks.empty())
+            {
+                std::thread shutdownCompletedThread{[self = shared_from_this()]()
+                                                    {
+                                                        self->waitForShutdown();
+                                                        vector<function<void()>> callbacks;
+                                                        {
+                                                            lock_guard lg(self->_mutex);
+                                                            callbacks = std::move(self->_shutdownCompletedCallbacks);
+                                                        }
+
+                                                        for (const auto& callback : callbacks)
+                                                        {
+                                                            callback();
+                                                        }
+                                                    }};
+
+                // Joining this thread is challenging because a callback can destroy the communicator and indirectly
+                // this object adapter factory. So we just detach it.
+                shutdownCompletedThread.detach();
+            }
+
+            _shutdownCompletedCallbacks.emplace_back(std::move(completed));
```

Comment:
I think std::push_back would be more idiomatic here since we're moving an existing function, not creating a new one.

Reply:
Fixed